File: Assignment2.rb
include 'rubygems'
include 'curb'
include 'FileTest'
require 'net/http'

require "open-uri"
require "fileutils"

def download(url, path)
  case io = open(url)
  when StringIO then File.open(path, 'w') { |f| f.write(io) }
  when Tempfile then io.close; FileUtils.mv(io.path, path)
  end
end

puts 'Author: Caolan O\'Domhnaill'
puts 'Date: Mon Oct 29 14:55:47 PDT 2018'
puts 'Assignment #2 - loads the robots.txt from many different domains and calculates some stats \
      The purpose of this script is to download the opendns top domains, and attempt to curl each of their robots.txt files.\
      Then it finds the average robots.txt file, and output the top 5 user-agents and disallowed'

url_top_domains = 'https://raw.githubusercontent.com'
path_to_domains_file = '/opendns/public-domain-lists/master/opendns-top-domains.txt'
top_domains_file = 'opendns-top-domains.txt'
file_size_col = 5
file_size_sum = 0
file_size_avg = 0
num_usable_domain_files = 0

list_of_outputs_files = [
  'user-agent.txt',
  'disallow.txt',
]

# remove the output files from the disk
for list_of_output_files.each do | filename |
  if File.file?(filename) then
    File.delete(filename)
  end
done

# Must be somedomain.net instead of somedomain.net/, otherwise, it will throw exception.
download(path_to_domains_file, top_domains_file)

File.open(file_top_domains).inject(0) { |num_domains, line| num_domains + 1 }

domain_count=0

puts "Looping through the contents of the file ${file_top_domains}"
for domain in $( cat ${file_top_domains} ); do
    
    full_robots_url = domain + String('/robots.txt')

    puts "Downloading the robots.txt from the domain ${domain}"
    curl -o robots.txt -L --connect-timeout 60 "${full_robots_url}"

    curl_result=$?
    if test ${curl_result} != "0"
    then
        puts "unable to download from ${domain}. Skipping..."
        continue
    fi
    
    # if file exists and is not empty
    puts "Checking to see if the robots.txt exists and is larger than 0 bytes"
        
    if File.file?('robots.txt') then
        puts "getting the size of the robots.txt file..."
        size = File.size('robots.txt')
        num_usable_domain_files += 1
        file_size_sum = file_size_sum + size
        
        grep -i "user-agent" robots.txt >> user-agent.txt
        grep -i "disallow" robots.txt >> disallow.txt
    else
        puts "robots.txt is empty for domain ${domain}"
    end
    
    # for testing purposes: limit the number of domains processed
    domain_count += 1
    
    if domain_count > 50 then
      break
    end
done

# Generate the average file size of robots.txt
avg = file_size_sum / num_usable_domain_files

puts "Average File Size: ${avg}"
puts "The Top Five user-agents: "
cat user-agent.txt | sort | tr -d '\r' | uniq -c | sort -u | sort -zn | tail -n 6
puts "The Top Five Disallows: "
cat disallow.txt | sort | tr -d '\r' | uniq -c | sort -u | sort -zn | tail -n 6
return 0

File: Assignment2.py
import urllib2
import socket
import time
from collections import defaultdict


"""
Function Name: download_file
Purpose: downloads a file and saves it to the same name
Inputs:
    - url: URL to where the file is on the net
    - filename: the file
Outputs: n/a
Notes: 
"""
def download_file(url, filename):
    num_attempts = 0
    max_attempts = 5
    f = None
    while num_attempts < max_attempts:
        try:
            full_url = "http://" + url + "/" + filename
            response = urllib2.urlopen(full_url, timeout = 5)
            content = response.read()
            f = open("./" + filename, 'w')
            f.write(content)
            f.close()
            break
        except socket.timeout as e:
            print type(e)
            print "There was an error: %r, on file: %s" % (e, full_url)
            break
        except urllib2.URLError as e:
            num_attempts += 1
        except Exception as e:
            break
        

"""
Function Name: output_top_five
Purpose: downloads a file and saves it to the same name
Inputs:
    - map_disallow: a map of the disallows and their counts
    - map_user_agents: a map of the user-agents and their counts
Outputs: n/a
Notes: 
"""    
def output_top_five(map_disallow, map_user_agents):
    sorted_disallowed = sorted(map_disallow.values())
    sorted_user_agents = sorted(map_user_agents.values())
    first5_disallowed = {i: sorted_disallowed[i] for i in sorted_disallowed.keys()[:5]}
    first5_user_agents = {i: sorted_user_agents[i] for i in sorted_user_agents.keys()[:5]}

    print "The top five disallowed are: "
    print sorted_disallowed.index

    print "The top five user-agents are: "
    print sorted_disallowed.index

"""
Function Name: read_domainlist_into_list
Purpose: Reads the domains into a list for iteration
Inputs:
    - filename: the name of the file to read in
Outputs:
    - a list of the domains
Notes: 
"""
def read_domainlist_into_list(filename):
    with open(filename) as f:
        content = f.readlines()
    return [item.strip() for item in content]

"""

"""
def build_maps(domain_list):
    disallow_map = defaultdict(int)
    useragent_map = defaultdict(int)
    robots_file = "robots.txt"
    
    print "Reading domain list..."
    for domain in domain_list:
        
        print "Processing http://%s/%s" % (domain, robots_file)
        download_file(domain, robots_file)
        with open(robots_file) as f:
            content = f.readlines()
        robots_lines = [item.strip() for item in content]
        for line in robots_lines:
            if "user-agent: " in line.lower():
                useragent_map[line] += 1
            elif "disallow: " in line.lower():
                disallow_map[line] += 1    
    return (disallow_map, useragent_map)

    
"""
Function Name: run
Purpose: execution happens in this scope
Inputs: n/a
Outputs: n/a
Notes: 
"""    
def run():

    robots_file = "robots.txt"
    url_top_domains = "raw.githubusercontent.com/opendns/public-domain-lists/master"
    file_top_domains = "opendns-top-domains.txt"
    
    # download the domain file and load into a list
    download_file(url_top_domains, file_top_domains)
    domains = read_domainlist_into_list(file_top_domains)
    
    start_time = time.time()
    (disallow, user_agents) = build_maps(domains)
    end_time = time.time()
    
    # Output the top items
    output_top_five(disallow, user_agents)
    
    
if __name__ == '__main__':
    run()